pylint/checkers/unicode.py

```diff
@@ -0,0 +1,670 @@
+        return codec.startswith("utf16") or codec.startswith("utf32")
+
+    def process_module(self, node: nodes.Module) -> None:
+        """
+        Perform the the actual checking by checking module stream
+
+        the module's content is accessible via node.stream() function
+        """
+        codec: Optional[str]
+        lines: Sequence[bytes]
+        invalid_codec_line: Optional[int] = None
+
+        # First try to detect encoding with PEP 263, won't work
+        # see https://bugs.python.org/issue1503789
+        # TO DO we could manually modify the stream to convert UTF16/UTF32 to ASCII
+        #      but why ;-) ?
+        try:
+            with node.stream() as stream:
+                codec, lines = detect_encoding(stream.readline)
+        except SyntaxError:
+            pass
+        else:
+            if self.is_invalid_codec(codec):
+                invalid_codec_line = len(lines)
+
+        # If there is no invalid codec already, lets try to determine a BOM,
+        # even so Python currently doesn't like non UTF-8 BOMs ....
+        if invalid_codec_line is None:
+            with node.stream() as stream:
+                codec = extract_codec_from_bom(stream.readline())
+            if self.is_invalid_codec(codec):
+                invalid_codec_line = 1
+
+        if invalid_codec_line is not None:
+            self.add_message(
+                "invalid-unicode-codec",
+                line=invalid_codec_line,
+                confidence=interfaces.HIGH,
+                col_offset=0,
+            )
+
+
+class BidirectionalUnicodeChecker(BaseChecker):
+    """
+    Check Unicode that could be used to hide bad code to humans
+    """
+
+    # Based on:
+    # https://golangexample.com/go-linter-which-checks-for-dangerous-unicode-character-sequences/
+    BAD_UNICODE = [
+        "\u202A",
+        "\u202B",
+        "\u202C",
+        "\u202D",
+        "\u202E",
+        "\u2066",
+        "\u2067",
+        "\u2068",
+        "\u2069",
+        # The following was part of the PEP 672 project:
+        # https://www.python.org/dev/peps/pep-0672/
+        # so the list above might not be complete
+        "\u200f",
+        # We don't use "\u200e" (LEFT to RIGHT MARK)
+        # as this is the default for latin files and can't be used
+        # to hide code
+    ]
+
+    __implements__ = interfaces.IRawChecker
+    priority = -1
+
+    name = "bidirectional_unicode_checker"
+    msgs = {
+        "E1801": (
+            "bidirectional unicode can be used to hide real code execution",
+            "bidirectional-unicode",
+            "bidirectional unicode can be used to hide real code execution",
+        )
+    }
+    options = ()
+
+    def process_module(self, node: nodes.Module) -> None:
+        """
+        Perform the the actual checking by checking module stream
+
+        the module's content is accessible via node.stream() function
+        """
+        with node.stream() as stream:
```

Comment:
Can't this be merged with the previous checker? Although somewhat dissimilar they are pretty small and use similar methods, so it might actually make sense performance-wise.

Reply:
Suggest a class name (hardest part) and I merge them

Reply:
Perhaps `UnicodeChecker` and for the other `AsciiChecker/ASCIIChecker`? I haven't looked at the third class yet so perhaps even one `UnicodeChecker` makes sense? I can't comment on that now though.

Reply:
Was a smart Idea to combine the Unicode checkers. This way I could use the codec determined and make Bidirectional Checker code ready for UTF-16/32.

Reply:
Combined them :-)